### DISCUSSION TITLE
Use devpi with Gitlab PyPi Repository

### DISCUSSION THREAD

**Discussion by lovetheguitar** [devpi/devpi#887] [Q&A]
Hi,

thanks for the great work on devpi first. 😃 

I'm trying to use devpi with our gitlab pypi registry. 
So far I tried adding the access token via a ~/.netrc file and directly in the url.

The problem appears to be related with devpi trying to call the pypi/simple api endpoint, where gitlab returns a 404 Not Found error.

I tried to add it like this `devpi index -c test/gitlab_mirror type=mirror mirror_url=https://mytoken:mytokenblabla@gitlab.lovethguitar.com/api/v4/groups/64/-/packages/pypi/simple/`

Here is the log output:
```bash
2022-04-14 19:41:35,300 INFO  [req14] GET /test/gitlab_mirror/+simple/
2022-04-14 19:41:35,301 DEBUG [req14] [Rtx146] in-transaction 146
2022-04-14 19:41:35,301 INFO  [req14] starting +simple
2022-04-14 19:41:35,303 DEBUG Starting new HTTPS connection (7): gitlab.lovetheguitar.com:443
2022-04-14 19:41:35,350 DEBUG https://gitlab.lovetheguitar.com:443 "GET /api/v4/groups/64/-/packages/pypi/simple/ HTTP/1.1" 404 25
2022-04-14 19:41:35,351 WARNI [req14] [Rtx146] upstream error (URL URL('https://gitlab.lovetheguitar.com/api/v4/groups/64/-/packages/pypi/simple/') returned 404 Not Found): using stale projects list
2022-04-14 19:41:35,352 DEBUG [req14] [Rtx146] closing transaction at 146
2022-04-14 19:41:35,352 DEBUG [req14] 200 OK 0.052s serial=146 length=None type=text/html; charset=UTF-8
```

Is there any setting I can tweak or any other way to combine root/pypi with our gitlab pypi registry to one index?

I can imagine that this would be a limitation or "bad" implementation on Gitlab's side? Either way, it would be very nice if it works somehow. 

Thanks a lot.

**Comment by fschulze** [comment]
Which devpi-server version do you use? Between >=6.2.0 and <6.5.0 there was a bug: https://github.com/devpi/devpi/issues/871

**Reply by lovetheguitar** [reply]
I'm using version 6.5.0 and I saw the issue, but I'm getting a 404 and not a 401 as in the issue.
I also tried to open the url (with and without token) in a browser and always got a 404. 😕 

Thanks for your help.

**Reply by fschulze** [reply]
After a quick glance at the API docs, I would say the ``/-`` in there is wrong:
https://docs.gitlab.com/ee/api/packages/pypi.html#group-level-simple-api-entry-point
It should be ``mirror_url=https://mytoken:mytokenblabla@gitlab.lovethguitar.com/api/v4/groups/64/packages/pypi/simple/``

**Reply by lovetheguitar** [reply]
Good catch, sadly doesn't seem to change anything.
I tried to open all the various urls in the browser and from python via `requests.get` and I can say that apparently the `.../simple/` endpoint always yields a 404 {"error":"404 Not Found"} response. 
No matter if I try to access the index on group level or project level, with or without the `/-`.

A normal pip install, specifying the gitlab index via `--index-url` with the urls above, works though.
I don't know exactly what pip does when installing a package, but I got the `/-` in the url from it's log. 
Here the pip log:

```bash
Looking in indexes: https://mytoken:****@gitlab.lovetheguitar.com/api/v4/groups/64/-/packages/pypi/simple/
1 location(s) to search for versions of testpackageabc:
* https://mytoken:****@gitlab.lovetheguitar.com/api/v4/groups/64/-/packages/pypi/simple/testpackageabc/
Fetching project page and analyzing links: https://mytoken:****@gitlab.lovetheguitar.com/api/v4/groups/64/-/packages/pypi/simple/testpackageabc/
Getting page https://mytoken:****@gitlab.lovetheguitar.com/api/v4/groups/64/-/packages/pypi/simple/testpackageabc/
Found credentials in url for gitlab.lovetheguitar.com
Starting new HTTPS connection (1): gitlab.lovetheguitar.com:443
https://gitlab.lovetheguitar.com:443 "GET /api/v4/groups/64/-/packages/pypi/simple/testpackageabc/ HTTP/1.1" 200 None
  Found link https://gitlab.lovetheguitar.com/api/v4/groups/64/-/packages/pypi/files/21f65a9ef4b04a3cd719a4c73cccb163be57c65e101e89bae36bb103964b8b20/testpackageabc-0.1.0-py3-none-any.whl#sha256=21f65a9ef4b04a3cd719a4c73cccb163be57c65e101e89bae36bb103964b8b20 (from https://gitlab.lovetheguitar.com/api/v4/groups/64/-/packages/pypi/simple/testpackageabc/) (requires-python:>=3.6.1,<4.0.0), version: 0.1.0
  ```

**Comment by fschulze** [comment]
I guess gitlab doesn't support the projects listing. pip looks directly at the project URL. We rely on the projects listing for information security reasons, as otherwise accessing private packages would leak their name to PyPI or in your case gitlab.

**Reply by lovetheguitar** [reply]
What a pity, that's what I was afraid of. 😕 

Is there some way around this if I fully trust my gitlab instance (self-hosted)? I assume currently not.

Thanks for your help. 🙂

**Reply by fschulze** [reply]
I would ask GitLab for the reason they don't support the listing. If they only have the uploaded packages in there, it shouldn't be expensive (performance wise) to implement.

**Reply by fschulze** [reply]
There is a for now undocumented ``mirror_no_project_list`` setting, but it seems like there are further problems: https://github.com/devpi/devpi/issues/1006
